src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala

```diff
@@ -52,11 +66,46 @@ object AnnexScalaInstance {
    * Using this cache and the Scala compiler cache, but disabling Zinc's classloader cache works, but doesn't seem to be
    * any different from leaving Zinc's cache enabled.
    */
-  def getAnnexScalaInstance(allJars: Array[File]): AnnexScalaInstance = {
-    val key = allJars.map(_.getCanonicalPath).mkString(":")
+  private def getAnnexScalaInstance(allJars: Array[File], workDir: Path): AnnexScalaInstance = {
+    // We need to remove the sandbox prefix from the paths in order to compare them.
+    val mapBuilder = Map.newBuilder[Path, Path]
+    allJars.foreach { jar =>
+      val comparableJarPath = jar.toPath().toAbsolutePath().normalize()
+      mapBuilder.addOne(jar.toPath -> workDir.toAbsolutePath().normalize().relativize(comparableJarPath))
+    }
+    val workRequestJarToWorkerJar = mapBuilder.result()
+
+    // Because we're just using file names here, theres's a problem  if the contents of the jars
+    // on the classpath for a particular version of the compiler change and we already have a
+    // ScalaInstance created and cached for the filenames on that classpath. In that case we'll use
+    // the cached ScalaInstance rather than create a new one.
+    //
+    // We could get around this problem by hashing all the jars and using that as the cache key,
+    // but that would require hashing the all the classpath jars on every compilation request. I
+    // imagine that would cause a performance hit.
+    //
+    // I also imagine it is extremeley rare to be mutating the contents of compiler classpath jars
+    // while keeping the names the same, e.g., generating new scala library jar for scala 2.13.14.
+    // As a result I'm leaving this string based for now.
+    val key = workRequestJarToWorkerJar.values.mkString(":")
 
     Option(instanceCache.get(key)).getOrElse {
-      val instance = new AnnexScalaInstance(allJars)
+      // Copy all the jars to the worker's directory because in a sandboxed world the
+      // jars can go away after the work request, so we can't rely on them sticking around.
+      // This should only happen once per compiler version, so it shouldn't happen often.
+      workRequestJarToWorkerJar.foreach { case (workRequestJar, workerJar) =>
+        this.synchronized {
+          // Check for existence of the file just in case another request is also writing these jars
```

Comment:
I think the race condition happens when you have two separate AnnexScalaInstances being created that share a jar.

I'm likely being overly cautious because the Javadoc states "Copying a file is not an atomic operation."

https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html

I was worried about calling `Files.copy`, it not being atomic, and having the same file being copied from two different places at once and ending up with some funky corruption or something like that.

Reply:
I suppose we're better safe than sorry. I just thought the locking done by `this.synchronized` prevented this sort of thing.